database/rrdset.c

```diff
@@ -519,18 +520,138 @@ RRDSET *rrdset_create_custom(
         if (!is_archived && rrdset_flag_check(st, RRDSET_FLAG_ARCHIVED)) {
             rrdset_flag_clear(st, RRDSET_FLAG_ARCHIVED);
+            mark_rebuild |= META_CHART_ACTIVATED;
         }
+        char *old_plugin = NULL, *old_module = NULL, *old_title = NULL, *old_family = NULL, *old_context = NULL;
+        const char *new_name = name ? name : id;
 
-        if(unlikely(name))
-            rrdset_set_name(st, name);
-        else
-            rrdset_set_name(st, id);
+        if (unlikely((st->name && !strcmp(st->name, new_name)) || !st->name)) {
+            mark_rebuild |= META_CHART_UPDATED;
+            rrdset_set_name(st, new_name);
+        }
+
+        if (unlikely(st->priority != priority)) {
+            st->priority = priority;
+            mark_rebuild |= META_CHART_UPDATED;
+        }
+        if (unlikely(st->rrd_memory_mode == RRD_MEMORY_MODE_DBENGINE && st->update_every != update_every)) {
+            st->update_every = update_every;
+            mark_rebuild |= META_CHART_UPDATED;
+        }
+
+        if (plugin && st->plugin_name) {
+            if (unlikely(strcmp(plugin, st->plugin_name))) {
+                old_plugin = st->plugin_name;
+                st->plugin_name = strdupz(plugin);
+                mark_rebuild |= META_PLUGIN_UPDATED;
+            }
+        } else {
+            if (plugin != st->plugin_name) { // one is NULL?
+                old_plugin = st->plugin_name;
+                st->plugin_name = plugin ? strdupz(plugin) : NULL;
+                mark_rebuild |= META_PLUGIN_UPDATED;
+            }
+        }
 
+        if (module && st->module_name) {
+            if (unlikely(strcmp(module, st->module_name))) {
+                old_module = st->module_name;
+                st->module_name = strdupz(module);
+                mark_rebuild |= META_MODULE_UPDATED;
+            }
+        } else {
+            if (module != st->module_name) {
+                if (st->module_name && *st->module_name) {
+                    old_module = st->module_name;
+                    st->module_name = module ? strdupz(module) : NULL;
+                    mark_rebuild |= META_MODULE_UPDATED;
+                }
+            }
+        }
+
+        char *new_title = title ? strdupz(title) : NULL;
+        if (new_title) {
+            json_fix_string(new_title);
+            if (unlikely(strcmp(st->title, new_title))) {
+                old_title = st->title;
+                st->title = new_title;
+                mark_rebuild |= META_CHART_UPDATED;
+            } else
+                freez(new_title);
+        }
+
+        RRDSET_TYPE new_chart_type =
+            rrdset_type_id(config_get(st->config_section, "chart type", rrdset_type_name(chart_type)));
+        if (st->chart_type != new_chart_type) {
+            st->chart_type = new_chart_type;
+            mark_rebuild |= META_CHART_UPDATED;
+        }
+
+        char *new_family = family ? strdup(family): NULL;
```

Comment:
This is the fast path and we shouldn't `strdup()` unconditionally, right? Also we should use `strdupz()`.